src/sidebar/components/button.js

```diff
@@ -0,0 +1,119 @@
+'use strict';
+
+const classnames = require('classnames');
+const propTypes = require('prop-types');
+const { createElement } = require('preact');
+
+const SvgIcon = require('./svg-icon');
+
+/**
+ * A button, which can be one of three core types depending on provided props:
+ * - icon-only (`icon` set, `label` missing)
+ * - text-only (`label` set, `icon` missing)
+ * - icon and text (`icon` and `label` present)
+ *
+ * In addition, a button may be adjusted:
+ * - `useCompactStyle`: to fit in a tighter spot
+ * - `useInputStyle`: to sit next to an input element
+ */
+function Button({
+  className = '',
+  icon = '',
+  isActive = false,
+  isPrimary = false,
+  label,
+  onClick = () => null,
+  title,
+  useCompactStyle = false,
+  useInputStyle = false,
+}) {
+  // If a `label` is provided, the `title` prop is optional and the `button`'s
+  // `title` attribute will be set from the `label`
+  title = title || label;
+
+  const baseClassName = label ? 'button--labeled' : 'button--icon-only';
+
+  return (
+    <button
+      className={classnames('button', baseClassName, className, {
+        'button--compact': useCompactStyle,
+        'button--input': useInputStyle,
+        'button--primary': isPrimary,
+        'is-active': isActive,
+      })}
+      onClick={onClick}
+      aria-pressed={isActive}
+      title={title}
+    >
+      {icon && <SvgIcon name={icon} className="button__icon" />}
+      {label && <span>{label}</span>}
```

Comment:
Which reminds me that @LMS007 finds the prop name `label` confusing because of the HTML `label` element...I'll tidy up.

Reply:
Hmmm, but there's gotta be a wrapping element of some sort here (unless I use `Fragment`, which seems a little silly).

Reply:
Why do you think you need a wrapping element? If you have a single string value (rather than a JSX element) it just generates a text node on its own, which is what we want. A `Fragment` is only needed when you want to have a context where you need a single JSX value (string, boolean or element), such as the right-hand side of a `{condition && jsxValue}` expression but have a group of values that you want to render.

Reply:
Because `{buttonText && ({buttonText})}` throws. Are you suggesting I render without checks, like `{buttonText}` alone? I am distrustful of p/react's JSX truthiness representation and worried about getting rendered `0`s, etc.

Reply:
> Because `{buttonText && ({buttonText})}` throws. 

Heh - Think about it 😉 : `{buttonText && ({buttonText})}` evaluates to an _object_ with a `buttonText` property due to the nested `{`s. Non-element objects are not legal JSX values.

> Are you suggesting I render without checks, like {buttonText} alone?

Yes. If the expression in `{` parens evaluates to a nullish or boolean value it will render nothing. This is guaranteed. Millions of lines of existing code depending on it!

Standalone demo:

```html
<body>
  <script src="https://unpkg.com/preact"></script>

  <div id="app"></div>

  <script>
    const { createElement, render } = preact;
    const appEl = document.querySelector('#app');

    function App() {
      return createElement('div', {},
        // Non-empty string value. This will render a text node.
        'first-string',
        // Empty string value. This renders nothing.
        '',
        // Number values. These will render text nodes.
        42, 0,
        // Null, undefined and booleans values. These all render nothing.
        null, false, undefined, true,
        // Element value. Renders a DOM element.
        createElement('p', {}, 'Child element')
      );
    }
    render(createElement(App), appEl);
  </script>
</body>
```